File: src/test/java/com/key/oa/event/LoginTest.java
package com.key.oa.event;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.key.oa.common.JsonResponse;
import com.key.oa.dto.LoginDTO;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import java.nio.charset.StandardCharsets;

@SpringBootTest
@AutoConfigureMockMvc
public class LoginTest {
    private final MockMvc mockMvc;

    private final ObjectMapper mapper;

    /**
     * 访问需要授权的URL时需要携带的token，放在header中
     * 注意！JUnit中测试方法无法共享非静态变量！
     */
    private static String token;

    @Autowired
    public LoginTest(MockMvc mockMvc, ObjectMapper objectMapper) {
        this.mockMvc = mockMvc;
        this.mapper = objectMapper;
    }

    @Test
    @Order(1)
    public void testLoginSuccess() throws Exception {
        LoginDTO loginDTO = new LoginDTO("20221390", "1234");

        mockMvc.perform(MockMvcRequestBuilders.post("/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(mapper.writeValueAsString(loginDTO)))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.code").value("00000"))
                .andDo(result -> {
                    String content = result.getResponse().getContentAsString();
                    token = mapper.readValue(content, new TypeReference<JsonResponse<String>>() {
                    }).getData();
                });
    }

    @Test
    public void testUsernameNotFound() throws Exception {
        LoginDTO loginDTO = new LoginDTO("12341234", "1234");

        mockMvc.perform(MockMvcRequestBuilders.post("/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(mapper.writeValueAsString(loginDTO))
                        .accept(MediaType.APPLICATION_JSON)
                        .characterEncoding(StandardCharsets.UTF_8))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.code").value("A0201"))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.message").value("用户账户不存在"))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.data").isEmpty());
    }

    @Test
    public void testPasswordWrong() throws Exception {
        LoginDTO loginDTO = new LoginDTO("20221390", "12341");

        mockMvc.perform(MockMvcRequestBuilders.post("/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(mapper.writeValueAsString(loginDTO))
                        .accept(MediaType.APPLICATION_JSON)
                        .characterEncoding(StandardCharsets.UTF_8))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.code").value("A0210"))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.message").value("用户密码错误"))
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.data").isEmpty());
    }

    @Test
    @Order(2)
    public void testLogoutSuccess() throws Exception {
        mockMvc.perform(MockMvcRequestBuilders
                        .post("/logout")
                        .header("token", token))
                .andExpect(MockMvcResultMatchers
                        .status()
                        .isOk())
                .andExpect(MockMvcResultMatchers
                        .jsonPath("$.code").value("00000"));
    }
}

File: src/test/kotlin/oa/event/LoginTest.kt
package oa.event

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import oa.common.JsonResponse
import oa.dto.LoginDTO
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.post

@SpringBootTest
@AutoConfigureMockMvc
class LoginTest @Autowired constructor(private val mockMvc: MockMvc, private val mapper: ObjectMapper) {
    private lateinit var token: String

    @Test
    @Order(1)
    @Throws(Exception::class)
    fun testLoginSuccess() {
        val loginDTO = LoginDTO("20221390", "1234")

        mockMvc.post("/login") {
            contentType = MediaType.APPLICATION_JSON
            content = mapper.writeValueAsString(loginDTO)
        }.andExpect {
            jsonPath("$.code") { value("00000") }
        }.andDo {
            handle {
                val content: String = it.response.contentAsString
                token = mapper.readValue(content,
                    object : TypeReference<JsonResponse<String>>() {}).data!!
            }
        }
    }

    @Test
    @Order(2)
    @Throws(java.lang.Exception::class)
    fun testLogoutSuccess() {
        mockMvc.post("/logout") {
            header("token", token)
        }.andExpect {
            status { isOk() }
            jsonPath("$.code") { value("00000") }
        }
    }

    @Test
    @Throws(java.lang.Exception::class)
    fun testUsernameNotFound() {
        val loginDTO = LoginDTO("12341234", "1234")
        mockMvc.post("/login") {
            contentType = MediaType.APPLICATION_JSON
            content = mapper.writeValueAsString(loginDTO)
            accept = MediaType.APPLICATION_JSON
            characterEncoding = "UTF-8"
        }.andExpect {
            jsonPath("$.code") { value("A0201") }
            jsonPath("$.message") { value("用户账户不存在") }
            jsonPath("$.data") { isEmpty() }
        }
    }

    @Test
    @Throws(java.lang.Exception::class)
    fun testPasswordWrong() {
        val loginDTO = LoginDTO("20221390", "12341")
        mockMvc.post("/login") {
            contentType = MediaType.APPLICATION_JSON
            content = mapper.writeValueAsString(loginDTO)
            accept = MediaType.APPLICATION_JSON
            characterEncoding = "UTF-8"
        }.andExpect {
            jsonPath("$.code") { value("A0210") }
            jsonPath("$.message") { value("用户密码错误") }
            jsonPath("$.data") { isEmpty() }
        }
    }
}